src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,547 @@
+		privdata->key_len = (env->algorithm_ref + 7)/8;
+		privdata->sec_env = (struct sc_security_env *)env;
+		privdata->se_num = se_num;
+		LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+	}
+
+	if (env->operation != SC_SEC_OPERATION_SIGN)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	if (!(env->flags & SC_SEC_ENV_KEY_REF_PRESENT))
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	if (env->key_ref_len != 1)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	if (env->algorithm != SC_ALGORITHM_EC)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x22, 0x81, 0xB6);
+	sbuf[0] = 0x91;
+	sbuf[1] = 0x02;
+	if (env->algorithm_flags & SC_ALGORITHM_ECDSA_HASH_SHA1)
+		sbuf[2] = 0x11;
+	else if (env->algorithm_flags & SC_ALGORITHM_ECDSA_HASH_SHA256)
+		sbuf[2] = 0x21;
+	else if (env->algorithm_flags & (SC_ALGORITHM_ECDSA_RAW|SC_ALGORITHM_ECDSA_HASH_NONE))
+		sbuf[2] = 0x22;
+	else
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	for (i = 0; i < MAX_OBJECTS && privdata->prkey_mappings[i][1]; i++) {
+		if (privdata->prkey_mappings[i][0] == env->key_ref[0])
+			break;
+	}
+	if (i == MAX_OBJECTS)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	sbuf[3] = privdata->prkey_mappings[i][1];
+	apdu.lc = 4;
+	apdu.datalen = 4;
+	apdu.data = sbuf;
+	if (se_num > 0) {
+		r = sc_lock(card);
+		LOG_TEST_RET(card->ctx, r, "sc_lock() failed");
+		locked = 1;
+	}
+	if (apdu.datalen) {
+		r = sc_transmit_apdu(card, &apdu);
+		if (r) {
+			sc_log(card->ctx, "%s: APDU transmit failed", sc_strerror(r));
+			goto err;
+		}
+		r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+		if (r) {
+			sc_log(card->ctx, "%s: Card returned error", sc_strerror(r));
+			goto err;
+		}
+	}
+	if (se_num <= 0) {
+		r = SC_SUCCESS;
+		goto err;
+	}
+	sc_unlock(card);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+
+	LOG_FUNC_RETURN(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2));
+err:
+	if (locked)
+		sc_unlock(card);
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+static int eoi_compute_signature(struct sc_card *card, const u8 * data, size_t data_len, u8 *out, size_t outlen)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	int r;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	/*
+	 * Guess the correct mode. If the size is less than the full-key-len, it must be a hash then
+	 */
+	if (privdata->key_len != data_len) {
+		switch (data_len) {
+			case SHA_DIGEST_LENGTH:
+				privdata->sec_env->algorithm_flags = SC_ALGORITHM_ECDSA_HASH_SHA1;
+				break;
+			case SHA256_DIGEST_LENGTH:
+				privdata->sec_env->algorithm_flags = SC_ALGORITHM_ECDSA_HASH_SHA256;
+				break;
+		}
+	}
+	/* Now we know which hash is used */
+	privdata->sec_env->algorithm_flags |= ALREADY_PROCESSED;
+
+	/* Perform the true set_security_env */
+	r = eoi_set_security_env(card, privdata->sec_env, privdata->se_num);
```

Comment:
I think we should set the `sec_env`  to NULL after using it here to make sure we do not use after it will go out of the scope, after returning from `sc_pkcs15_compute_signature()` where it is defined.

Even better might be to copy the structure in the `eoi_set_security_env()` to make sure it will not get out of scope when somebody would get to call this function weirdly.

Reply:
Done. I now use a local copy. I've noticed that `sc_security_env.params[].value` is of type `void*`. So I guess, it's not exactly a full copy, but I don't think the `value` is used in this case, so I guess it's ok.